Avoid routing 2q barriers in SabreSwap

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit fixes an oversight in the sabre swap pass where a 2 qubit barrier would have been treated like a 2 qubit gate and swaps could potentially be inserted if the sabre algorithm thought it didn't have connectivity for the qargs to the barrier. However as barrier is just a compiler directive it is valid on all qubit pairs so this swap insertion was unnecessary, it was still valid but just not an optimal output. This commit fixes it by adding context to the sabre dag around whether a given node is a directive (and valid on all qubits) or not.


### Details and comments

mtreinish committed:
Avoid routing 2q barriers in SabreSwap

This commit fixes an oversight in the sabre swap pass where a 2 qubit
barrier would have been treated like a 2 qubit gate and swaps could
potentially be inserted if the sabre algorithm thought it didn't have
connectivity for the qargs to the barrier. However as barrier is just
a compiler directive it is valid on all qubit pairs so this swap
insertion was unnecessary, it was still valid but just not an optimal
output. This commit fixes it by adding context to the sabre dag around
whether a given node is a directive (and valid on all qubits) or not.

mtreinish committed:
Update rust tests too

kevinhartman reviewed (commented):
I believe you also need to update extended set population to ignore directives:

```rs
if required_predecessors[successor_index] == 0 {
    if !dag.dag[node].directive && !dag.node_blocks.contains_key(&successor_index) {
        if let [a, b] = dag.dag[successor_node].qubits[..] {
            extended_set.push([a.to_phys(layout), b.to_phys(layout)]);
        }
    }
    to_visit.push(successor_node);
}
```

Otherwise, this looks good!

mtreinish committed:
Skip directives in extended set generation too

Co-authored-by: Kevin Hartman <kevin@hart.mn>

mtreinish committed:
Merge branch 'main' into sabre-no-barrier

mtreinish commented:
> I believe you also need to update extended set population to ignore directives:

Done in: https://github.com/Qiskit/qiskit/pull/11295/commits/89da56ea20f8fc6b90ad215a32a89f0b34042f98

mtreinish committed:
Add release note

mtreinish committed:
Update test layout in test_layout_many_search_trials

This commit updates a layout in the:
`TestSabreLayout.test_layout_many_search_trials` test case which was
recently changed in #10323. The change in test ouput caused by #10323
is what triggered the investigation into this bugfix and now that
barriers are being treated correctly by sabre the layout doesn't change
in the test case anymore and this is reverting the test assertion to use
the original layout before #10323 was merged.

kevinhartman reviewed (approved):
LGTM, thanks for the updates!
